drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -0,0 +1,660 @@
+}
+
+/**
+ * @brief Set number of active time slots.
+ * @param dev - Device handler.
+ * @param timeslot_no - Last time slot to be enabled.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_last_timeslot(struct adpd410x_dev *dev,
+				   enum adpd410x_timeslots timeslot_no)
+{
+	int32_t ret;
+	uint16_t data;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_OPMODE, &data);
+	if(ret != SUCCESS)
+		return ret;
+	data &= ~BITM_OPMODE_TIMESLOT_EN;
+	data |= ((timeslot_no & 0x0f) << BITP_OPMODE_TIMESLOT_EN);
+
+	return adpd410x_reg_write(dev, ADPD410X_REG_OPMODE, data);
+}
+
+/**
+ * @brief Set device sampling frequency.
+ * @param dev - Device handler.
+ * @param sampling_freq - New sampling frequency.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_set_sampling_freq(struct adpd410x_dev *dev,
+				   uint32_t sampling_freq)
+{
+	int32_t ret;
+	uint32_t reg_load;
+	uint16_t reg_temp;
+
+	if((dev->clk_opt == ADPD410X_INTLFO_INTHFO) ||
+	    (dev->clk_opt == ADPD410X_INTLFO_EXTHFO)) {
+		ret = adpd410x_reg_read(dev, ADPD410X_REG_SYS_CTL, &reg_temp);
+		if(ret != SUCCESS)
+			return FAILURE;
+		if(reg_temp & BITP_SYS_CTL_LFOSC_SEL)
+			reg_load = ADPD410X_LOW_FREQ_OSCILLATOR_FREQ1;
+		else
+			reg_load = ADPD410X_LOW_FREQ_OSCILLATOR_FREQ2;
+	} else {
+		reg_load = dev->ext_lfo_freq;
+	}
+
+	reg_load /= sampling_freq;
+	ret = adpd410x_reg_write(dev, ADPD410X_REG_TS_FREQ,
+				 (reg_load & 0xFFFF));
+	if(ret != SUCCESS)
+		return FAILURE;
+	return adpd410x_reg_write(dev, ADPD410X_REG_TS_FREQH,
+				  ((reg_load & 0x7F0000) >> 16));
+}
+
+/**
+ * @brief Setup an active time slot.
+ * @param dev - Device handler.
+ * @param timeslot_no - Time slot ID to setup.
+ * @param init - Pointer to the time slot initialization structure.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_timeslot_setup(struct adpd410x_dev *dev,
+				enum adpd410x_timeslots timeslot_no,
+				struct adpd410x_timeslot_init *init)
+{
+	int32_t ret;
+	uint16_t data;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_OPMODE, &data);
+	if(ret != SUCCESS)
+		return ret;
+	if(((data & BITM_OPMODE_TIMESLOT_EN) >> BITP_OPMODE_TIMESLOT_EN) <
+	    timeslot_no)
+		return FAILURE;
+
+	/* Enable channel 2 */
+	ret = adpd410x_reg_read(dev, (ADPD410X_REG_TS_CTRL_A + timeslot_no * 0x20),
+				&data);
+	if(ret != SUCCESS)
+		return FAILURE;
+	if(init->enable_ch2)
+		data |= BITM_TS_CTRL_A_CH2_EN_A;
+	else
+		data &= ~BITM_TS_CTRL_A_CH2_EN_A;
+	ret = adpd410x_reg_write(dev, (ADPD410X_REG_TS_CTRL_A + timeslot_no * 0x20),
+				 data);
+	if(ret != SUCCESS)
+		return FAILURE;
+
+	/* Setup inputs */
+	data = init->ts_inputs.option << (init->ts_inputs.pair * 4);
+	ret = adpd410x_reg_write(dev, (ADPD410X_REG_INPUTS_A + timeslot_no * 0x20),
+				 data);
+	if(ret != SUCCESS)
+		return FAILURE;
+
+	/* Set precondition PD */
+	ret = adpd410x_reg_read(dev, (ADPD410X_REG_CATHODE_A + timeslot_no * 0x20),
```

Comment:
There are many occurences of this, it looks like timeslot_no * 0x20 could also be a register macro that takes the timeslot_no as argument.